User config watcher

asymmetric opened the pull request:
This PR adds a watcher for the `user.toml` file.

It's based on @krnowak's still-unmerged work on the file watcher.

Commit:
Add UserConfigWatcher

The UserConfigWatcher watches each service's user.toml file.
It reacts on file creation, deletion and change, triggering a service
reload and configuration.

If a service has hooks defined, those are executed, otherwise the
default actions are performed instead.

Commit:
Use a constant for user.toml

It's used in several places, so a constant makes sense.

Commit:
Improve documentation

Commit:
Remove incorrect comment for function

The comment was introduced in a9c98dd435c278c6dda0861cc27f3e900e54d8c7
and was incorrect back then as well. In the meantime, the function has been
changed and now doesn't return any value.

asymmetric commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -660,6 +662,7 @@ impl Service {
     }
 
     /// Write service files from gossip data to disk.
+    /// For the location, check `svc_file_path()`.
```
Wasn't sure how to create a symlink for this. Also, the function is private, so I'm not sure if there's a point.

krnowak replied:
I guess this is a note to yourself, because I don't understand what it is about. Not enough context.

asymmetric replied:
The `svc_file_path` function could be rendered as a link, since `///` comments are Markdown. I just don't know what the target should be, as it seems to be relative to something.

asymmetric replied:
Fixed this.

asymmetric commented:
**NB**: The config is not currently reloaded, so this doesn't work yet.

krnowak commented on components/sup/src/manager/mod.rs:
```diff
@@ -277,12 +280,13 @@ impl Manager {
             service_states: HashMap::new(),
             sys: Arc::new(sys),
             peer_watcher: peer_watcher,
+            user_config_watcher: UserConfigWatcher::new(),
```
If you changed the order of members in `Manager` struct, it would be good to reorder the initializations here too.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+
+
+pub struct UserConfigWatcher {
+    // TODO use a HashSet? we don't really need the boolean:
```
This comment does not seem to be true - `reset_events_for` sets the boolean to false.

asymmetric replied:
Yes, this is a note to myself saying that we could change the implementation. True/false values can be represented by presence/absence in a set.

asymmetric replied:
Tried using a `HashSet` and turns out it's more complicated than expected (if we put a `HashSet` in an `Arc`, that `HashSet` is immutable, unless I use mutexes), so not worth it.

krnowak replied:
Yeah, you would need to do something similar to what some code in butterfly does - `Arc<RwLock<HashSet<Whatever>>>`. Anyway, I think that current solution is fine as well.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+                );
+                let callbacks = UserConfigCallbacks { have_events: have_events };
+                let mut file_watcher =
+                    default_file_watcher(&path, callbacks).expect("creating user-config watcher");
```
Instead of using `expect`, I'd use `outputln!` macro to report an error and maybe quit the thread. I'm not sure if we should panic in this case.

asymmetric replied:
Good point, I've committed a change.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+        File::create(&file_path).expect("creating file");
+
+        // Allow the watcher to notice that a file was created.
+        thread::sleep(Duration::from_millis(100));
```
Why sleeping instead of `wait_for_events`?

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+use hcore::service::ServiceGroup;
+use manager::service::Service;
+
+static LOGKEY: &'static str = "UCM";
```
Consider adding this log key to the documentation here - `www/source/docs/reference/log-keys.html.md`.

krnowak reviewed (changes requested):
Looks good overall, I had some nits.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+                name: String::from("foo"),
+                path: TempDir::new("user-config-watcher")
+                    .expect("creating temp dir")
+                    .into_path(),
```
One note about this - when you use the `into_path()` function, you also take over the responsibility of removing the created temporary directory. Which you probably don't do, no?

asymmetric replied:
Good point, fixed.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,251 @@
+// This trait exists to ease the testing of functions that receive a Service.  Creating Services
+// requires a lot of ceremony, so we work around this with this trait.
+pub trait Serviceable {
+    fn name(&self) -> &String;
```
Returning references makes the trait less flexible, because you need an existing, non-temporary thing in memory to return its address (like a member of a struct that implements this trait). In other words - you can't compute the return value. Not sure if it is really important, but in general I would just return plain values. Of course the downside of returning values is that you need to clone the stuff, but OTOH this won't likely be a bottleneck ever.

asymmetric replied:
My rationale here is that the name is always a reference to its containing struct.

Isn't it enough for users of this trait to be able to call `name.clone()` if they want to own the `String`?

Can you show an example of where the current implementation creates issues? As I'm not sure I understand your comment.

krnowak replied:
There are no problems with current implementation. It was just a general note about returning references in trait functions. So in this case it's likely ok, but in general I would avoid it.

asymmetric replied:
OK, let me know if you feel strongly about changing this, otherwise I'm gonna leave it as is.

krnowak replied:
Nah, leave it as is, please. If somebody else points out this thing, then you can change it.

krnowak reviewed (changes requested):
Another two things I noticed.

Commit:
Set reload flag if user_config_updated is set

`user_config_updated` is set in the manager's
`update_running_services_from_user_config_watcher` function, which
receives events from the user-config watcher.

This ensures that the service is then reloaded whenever the
configuration is updated.

asymmetric commented:
@krnowak addressed your comments, but I guess this will just wait until your PR gets merged.

I also rebased on yours.

I will clean up the commit messages next week.

krnowak commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -387,29 +392,33 @@ impl Service {
             let (reload, reconfigure) = {
                 let ctx = self.render_context(census_ring);
-                let reload = self.compile_hooks(&ctx);
+                let reload = self.compile_hooks(&ctx) || self.user_config_updated;
```
Should this `|| self.user_config_updated` be moved one line below, to the `let reconfigure` line?

asymmetric replied:
No, the `needs_reconfiguration` flag is only used to [trigger a `reconfigure` hook](https://github.com/habitat-sh/habitat/blob/c827ae61fa9077974e22fdbf04aecbafefbcef19/components/sup/src/manager/service/mod.rs#L639-L646), which may not be defined. The `reload` flag is actually going to trigger a service reload, which is what we need here.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,260 @@
+        Self { states: HashMap::new() }
+    }
+
+    pub fn add<T: Serviceable>(&mut self, service: &T) {
```
We have an `add` function. Do we need a `remove` function then?

krnowak reviewed (commented):
I had two questions, otherwise looks good.

asymmetric commented:
The PR is in a weird state right now. It shows commit that are not in the PR's branch.

Also, please don't be to picky about commit messages (and the commits themselves), as I am still testing this PR out.

Commit:
Set cfg_updated to true if user_config_updated is

The `cfg_updated` variable is used to determine whether to run the
following block, in which we rebuild the configuration.

We need this to happen if the user-config watcher has detected changes to the
user-config file.

Commit:
Reset user_config_updated after using it

After we have performed all the actions required on detecting that the
user-config has changed, we revert the field to `false`, so that we
don't perform them again unless there are more changes.

Commit:
Call load_user when the configuration has changed

The `load_user` function is in charge of loading the user-config file.
We need this function to be called every time the user-config watcher
has detected change to that file.

Commit:
Stop watcher thread when service is removed

By adding another channel, we allow the User Config Watcher to
communicate to any of the Workers that they should stop running.

This allows the manager to cleanly remove a service.

Commit:
Initialize UserConfigWatcher in Manager

Commit:
Rename watcher field

Now that there are multiple watcher, it makes sense to make the SpecWatcher field more descriptive

Commit:
Group watcher fields

Commit:
Document User-Config Watcher logging key

Commit:
Add service to Watcher

Commit:
Update running services from UCW on each tick

Commit:
Document change in reconfigure hook

Commit:
Use SyncChannel for events

Using a sync channel with buffer size of 1 allows us to mimic the
behaviour of a boolean. We are only interested in the fact that there
were events, not how many (since the message on this channel is always
`()`).

Commit:
Extract callback implementation

Commit:
Use _ when matching channel messages

We are not interested in the messages themselves, only in their presence

krnowak committed:
Allow skipping the initial event in file watcher

Signed-off-by: Krzesimir Nowak <krzesimir@kinvolk.io>

Commit:
Don't run compile_hooks when user.toml changes

Before this refactor, the "Hooks recompiled" message was being printed
twice:

The first time, because the `user.toml` had been updated, which in turn meant the
service got restarted.
That counted as a change in the census ring, which in turn caused the
hooks to be recompiled and the message to be printed again.

With this refactor, we avoid running compile_hooks when the only thing
that's changed is the user config.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Commit:
Only load user.toml when needed

Before this change, we were calling the `load_user` method every time
there was a change, even if it wasn't unrelated to user config.

Now, we only do it if the `user_config_updated` flag is set.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Commit:
Add default_file_watcher_with_no_initial_event

This function mirrors `default_file_watcher`, but instead of calling
create it calls `create_with_no_initial_event`.

It is used to start a default file watcher, where the first event
triggered by the file already being there isn't emitted.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Commit:
Ignore initial event in UserConfigWatcher

Before this change, if the user.toml file was already present, the
service would be restarted immediately after being booted, because the
File Watcher would notice the file and emit an event.

With this function, the File Watcher doesn't emit the event anymore.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

asymmetric commented:
@krnowak could you take a look again?

krnowak commented on components/sup/src/manager/mod.rs:
```diff
@@ -879,20 +904,30 @@ impl Manager {
+        // The problem we're trying to work around here by adding this block is that `write`
+        // creates an immutable borrow on `self`, and `self.remove_service` needs `&mut self`.
+        // The solution is to introduce the block to drop the borrow before the call to
+        // `self.remove_service`, and use `mem::swap` to copy the services to a variable defined
```
To move, not copy. After swapping, the vector under `self.services` lock will be empty.

krnowak commented on components/sup/src/manager/service/mod.rs:
```diff
@@ -404,29 +409,49 @@ impl Service {
+        let cfg_changed = cfg_updated_from_rumors || self.user_config_updated;
+        let something_to_do = cfg_changed || census_ring.changed();
+
+        if something_to_do {
```
This became a bit hard to follow. Would pseudo-rust below be clearer? (`cufr` = `cfg_updated_from_rumors`, `sucu` = `self.user_config_updated`, `crc` = `census_ring.changed()`)

```
if sucu {
    if let Err(e) = self.cfg.load_user(&self.pkg) {…}
}
let (cfg_changed, reload, reconfigure) = {
    let ctx = …;
    if cufr || crc {
        (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))
    } else if sucu {
        (true, true, self.compile_configuration(&ctx))
    } else {
        (false, self.needs_reload, self.needs_reconfiguration)
    }
}
self.needs_reload = reload;
self.needs_reconfiguration = reconfigure;
self.user_config_updated = false;
/* return */ cfg_changed
```

krnowak replied:
The `something_to_do` variable would be gone.

asymmetric replied:
I find the 3 value tuple even less clear tbh...

> (cufr || sucu, self.compile_hooks(&ctx) || sucu, self.compile_configuration(&ctx))

what?

krnowak replied:
Not sure what this "what?" is about, but to be clear - I don't actually want to have variables like `cufr` or `sucu`. I wrote it like that to avoid typing in the comment.

asymmetric replied:
I think that even with more descriptive names, it's hard to parse logically. There's just too much happening on one line: returning a tuple, ORing two values, calling functions..

Not sure it's an improvement.

krnowak replied:
You don't need to put everything in one line. You could very well have:
```
let config_changed = config_updated_from_rumors || self.user_config_updated;
let needs_reload = self.compile_hooks(&ctx) || self.user_config_updated;
let needs_reconfigure = self.compile_configuration(&ctx);
(config_changed, needs_reload, needs_reconfigure)
```

In general, what I wanted to propose is to simplify the flow of the code, instead of using nested `if`s and `match`es that are sometimes executed and sometimes not.

asymmetric replied:
In the above implementation though, we're calling `self.compile_hooks` even if we don't need to, which will lead to "Hooks recompiled" being printed over and over (this function gets executed [on every tick](https://github.com/kinvolk/habitat/blob/0a1d210328174ae822c13213e2462f6f37f1e94c/components/sup/src/manager/service/mod.rs#L294)).

I'm definitely open to simplifying the code but obviously it has to have the same behaviour.

The problem is that some of these functions have side effects so we can't just call them willy-nilly.

krnowak replied:
The `self.compile_hooks(&ctx)` should only be executed when either `cfg_updated_from_rumors` is true or if `census_ring.changed()` returns true. So if this happens everytime, then something is wrong.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,338 @@
+    stop_running: Sender<()>,
+}
+
+type ServiceName = String;
```
That's nice.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,338 @@
+        Ok(())
+    }
+
+    /// Removes a service from the User Config Watcher, thereby stopping the watcher thread.
```
I'd add "eventually" - "thereby eventually stopping …"

asymmetric replied:
Why eventually? Do you mean that the operation is asynchronous?

If so, I think I can explain it more explicitly.

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,338 @@
+        Ok(())
+    }
+
+    /// Checks whether the watcher for the specified service has observed any events, thereby
```
thereby is your new favourite word, eh? :)

asymmetric replied:
It's a pretty cool word, you gotta admit that

krnowak commented on components/sup/src/manager/user_config_watcher.rs:
```diff
@@ -0,0 +1,338 @@
+    /// consuming them.
+    pub fn have_events_for<T: Serviceable>(&self, service: &T) -> bool {
+        if let Some(state) = self.states.get(service.name()) {
+
```
Drop this empty line?

krnowak reviewed (changes requested):
Looks mostly fine, but there was one hairy piece of code I'd like to see if it can be "balder".

Commit:
Improve comments

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Commit:
Improve debug message when thread dies

The thread is currently not restarted.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

Commit:
Refactor update_templates method

This refactor aims at simplifying the flow of the code and reducing
conditionals.

Signed-off-by: Lorenzo Manacorda <lorenzo@kinvolk.io>

asymmetric commented:
Closing in favor of https://github.com/habitat-sh/habitat/pull/3898.
